src/server/game/Handlers/AuctionHouseHandler.cpp

```diff
@@ -62,725 +58,171 @@ void WorldSession::SendAuctionHello(ObjectGuid guid, Creature* unit)
-        stmt->SetData(0, auction->bidder.GetCounter());
-        stmt->SetData(1, auction->bid);
-        stmt->SetData(2, auction->Id);
-        trans->Append(stmt);
-
-        SendAuctionCommandResult(auction->Id, AUCTION_PLACE_BID, ERR_AUCTION_OK, 0);
-    }
-    else
-    {
-        //buyout:
-        if (player->GetGUID() == auction->bidder)
-            player->ModifyMoney(-int32(auction->buyout - auction->bid));
-        else
-        {
-            player->ModifyMoney(-int32(auction->buyout));
-            if (auction->bidder)                          //buyout for bidded auction ..
-                sAuctionMgr->SendAuctionOutbiddedMail(auction, auction->buyout, GetPlayer(), trans);
-        }
-        auction->bidder = player->GetGUID();
-        auction->bid = auction->buyout;
-        GetPlayer()->UpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_HIGHEST_AUCTION_BID, auction->buyout);
-
-        //- Mails must be under transaction control too to prevent data loss
-        sAuctionMgr->SendAuctionSalePendingMail(auction, trans);
-        sAuctionMgr->SendAuctionSuccessfulMail(auction, trans);
-        sAuctionMgr->SendAuctionWonMail(auction, trans);
-        sScriptMgr->OnAuctionSuccessful(auctionHouse, auction);
-
-        SendAuctionCommandResult(auction->Id, AUCTION_PLACE_BID, ERR_AUCTION_OK);
-
-        auction->DeleteFromDB(trans);
-
-        sAuctionMgr->RemoveAItem(auction->item_guid);
-        auctionHouse->RemoveAuction(auction);
-    }
-    player->SaveInventoryAndGoldToDB(trans);
-    CharacterDatabase.CommitTransaction(trans);
+    sAsyncAuctionMgr->PlaceBid(_player->GetGUID(), packet.Auctioneer, packet.AuctionID, packet.Price);
 }
 
-//this void is called when auction_owner cancels his auction
-void WorldSession::HandleAuctionRemoveItem(WorldPacket& recvData)
+// Called when auction_owner cancels his auction
+void WorldSession::HandleAuctionRemoveItem(WorldPackets::AuctionHouse::RemoveItem& packet)
 {
     LOG_DEBUG("network", "WORLD: Received CMSG_AUCTION_REMOVE_ITEM");
 
-    ObjectGuid auctioneer;
-    uint32 auctionId;
-    recvData >> auctioneer;
-    recvData >> auctionId;
-
-    Creature* creature = GetPlayer()->GetNPCIfCanInteractWith(auctioneer, UNIT_NPC_FLAG_AUCTIONEER);
+    Creature* creature = GetPlayer()->GetNPCIfCanInteractWith(packet.Auctioneer, UNIT_NPC_FLAG_AUCTIONEER);
     if (!creature)
     {
-        LOG_DEBUG("network", "WORLD: HandleAuctionRemoveItem - Unit ({}) not found or you can't interact with him.", auctioneer.ToString());
+        LOG_DEBUG("network", "WORLD: HandleAuctionRemoveItem - Unit ({}) not found or you can't interact with him.", packet.Auctioneer.ToString());
         return;
     }
 
-    // remove fake death
+    // Remove fake death
     if (GetPlayer()->HasUnitState(UNIT_STATE_DIED))
         GetPlayer()->RemoveAurasByType(SPELL_AURA_FEIGN_DEATH);
 
     AuctionHouseObject* auctionHouse = sAuctionMgr->GetAuctionsMap(creature->GetFaction());
```

Comment:
Since ActionMgr is used in several threads I think it's unsafe to use it like this without synchronising mechanism.

Reply:
1 async thread, no more

Reply:
So the first thread - async thread with auction tasks, and the second one (as we can see from snippet above) is thread that processes opcodes (in this case HandleAuctionRemoveItem). So both threads seem to be using the same sAuctionMgr object without synchronization primitives. This could lead to race conditions. Is there something that I am missing or overlooking?